src/client/action/list_databases.rs

```diff
@@ -0,0 +1,168 @@
+use std::{future::IntoFuture, marker::PhantomData};
+
+use bson::{Bson, Document};
+use futures_util::FutureExt;
+
+#[cfg(any(feature = "sync", feature = "tokio-sync"))]
+use crate::sync::Client as SyncClient;
+use crate::{
+    client::BoxFuture,
+    error::{ErrorKind, Result},
+    operation::list_databases as op,
+    results::DatabaseSpecification,
+    Client,
+    ClientSession,
+};
+
+impl Client {
+    /// Gets information about each database present in the cluster the Client is connected to.
+    pub fn list_databases(&self) -> ListDatabases {
+        ListDatabases {
+            client: &self,
+            options: Default::default(),
+            session: None,
+            mode: PhantomData,
+        }
+    }
+
+    /// Gets the names of the databases present in the cluster the Client is connected to.
+    pub fn list_database_names(&self) -> ListDatabases<'_, Names> {
+        ListDatabases {
+            client: &self,
+            options: Default::default(),
+            session: None,
+            mode: PhantomData,
+        }
+    }
+}
+
+#[cfg(any(feature = "sync", feature = "tokio-sync"))]
+impl SyncClient {
+    /// Gets information about each database present in the cluster the Client is connected to.
+    pub fn list_databases(&self) -> ListDatabases {
+        self.async_client.list_databases()
+    }
+}
+
+/// Gets information about each database present in the cluster the Client is connected to.
+#[must_use]
+pub struct ListDatabases<'a, M: Mode = Full> {
+    client: &'a Client,
+    options: Option<op::Options>,
+    session: Option<&'a mut ClientSession>,
+    mode: PhantomData<M>,
+}
+
+mod private {
+    pub trait Sealed {}
+    impl Sealed for super::Full {}
+    impl Sealed for super::Names {}
+}
+
+pub trait Mode: private::Sealed {}
+impl Mode for Full {}
+impl Mode for Names {}
+
+pub struct Full;
+pub struct Names;
+
+impl<'a, M: Mode> ListDatabases<'a, M> {
+    fn options(&mut self) -> &mut op::Options {
+        self.options.get_or_insert_with(op::Options::default)
+    }
+
+    #[cfg(test)]
+    pub(crate) fn with_options(mut self, value: impl Into<Option<op::Options>>) -> Self {
+        self.options = value.into();
+        self
+    }
+
+    /// Filters the query.
+    pub fn filter(mut self, value: impl Into<Option<Document>>) -> Self {
+        self.options().filter = value.into();
+        self
+    }
+
+    /// Determines which databases to return based on the user's access privileges. This option is
+    /// only supported on server versions 4.0.5+.
+    pub fn authorized_databases(mut self, value: impl Into<Option<bool>>) -> Self {
```

Comment:
I'd rather have these values be the simple types to keep things clean and keep compile time down, but that makes it a hassle for any calling code to have optional values for these.

Reply:
Is there a real use case for passing `None` into any of these setter methods? For most (if not all) of our options `None` is the default and means "don't send anything," so I'm not sure why users would need to do so. If not I think we should consider taking only non-optional types instead to simplify the API. Maybe that's something we could flesh out during the pre-release period.

Reply:
The extremely hand-wavy hypothetical is when the client is writing something where they want to expose the optional nature of a setting in their own API.  In that case, if we're not accepting an `Option` they'd have to go through an awkward little dance:
```
let user_auth_param: Option<bool> = ...;// something external
let mut act = client.list_databases();
if let Some(auth) = user_auth_param {
  act = act.authorized_databases(auth);
}
let result = act.await;
```
rather than just
```
let user_auth_param: Option<bool> = ...;// something external
let result = client.list_databases().authorized_databases(user_auth_param).await;
```

Given that (a) this seems pretty unlikely and (b) it's still doable if needed, I think you're right and I'll go with the simple option and we can see if this comes up in pre-release.